sdm.R

```diff
@@ -0,0 +1,147 @@
+# From forecasting-bbs-R.ipynb --------------------------------------------
+sapply(dir("R", full.names = TRUE), source)
+start_yr <- 1982
+end_yr <- 2013
+min_num_yrs <- 25
+
+# start SDM ---------------------------------------------------------------
+library(gbm)
+library(parallel)
+library(dplyr)
+library(tidyr)
+
+set.seed(0)
+mc.cores = 16
+n_test_years = 5
+interaction.depth = 8
+n.trees = 1E4
+
+train_years = start_yr:(end_yr - n_test_years)
+test_years = (end_yr - n_test_years + 1):end_yr
+
+# Parameters for spatial cross-validation
+n_clusters = 100
+n_folds = 10
+stopifnot(n_clusters %% n_folds == 0)
+
+# Load data ---------------------------------------------------------------
+occurrences = get_bbs_data(start_yr = 1982, end_yr = 2013, min_num_yrs = 25)
+
+env = get_env_data() %>%
+  filter_ts(start_yr, end_yr, min_num_yrs) %>%
+  inner_join(distinct(occurrences, site_id, lat, long, year),
+             by = c("site_id", "year"))
+
+# Determine presence-absence ----------------------------------------------
+
+x = tidyr::expand(occurrences, site_id, year) %>%
+  inner_join(env, by = c("site_id", "year")) %>%
+  arrange(year, site_id)
+
+locations = distinct(x, site_id, lat, long)
+
+y = distinct(occurrences, year, species_id, site_id, .keep_all = TRUE) %>%
+  inner_join(select(x, year, site_id), c("year", "site_id")) %>%
+  mutate(presence = abundance > 0) %>%
+  select(-abundance, lat, long, start_time, month, day) %>%
+  spread(key = species_id, value = presence, fill = 0) %>%
+  arrange(year, site_id)
+
+
+# Spatial cross validation ------------------------------------------------
+# assign many clusters of routes that are close in lat/long
+cluster_id = kmeans(locations[ , 2:3], n_clusters, nstart = 100)$cluster
+
+# Randomly combine the clusters to make the cross-validation folds
+fold_matrix = matrix(sample(1:n_clusters), ncol = n_folds)
+fold_id = numeric(nrow(locations))
+for (i in 1:n_folds) {
+  fold_id = fold_id + i * (cluster_id %in% fold_matrix[, i])
+}
+
+x = x %>%
+  inner_join(
+    data_frame(site_id = locations$site_id, fold_id = fold_id),
+    "site_id"
+  )
+
+stopifnot(x$site_id == y$site_id, x$year == y$year)
+
+train_x = filter(x, year %in% train_years)
+test_x = filter(x, !(year %in% train_years))
+train_y = filter(y, year %in% train_years)
+test_y = filter(y, !(year %in% train_years))
+
+fold_ids = train_x$fold_id  #plot(locations[,3:2], col = fold_ids, pch = fold_id)
+
+cv_train_rows = function(i){
+  i != fold_ids
+}
+cv_test_rows = function(i){
+  i == fold_ids
+}
+# fit gbm -----------------------------------------------------------------
+
+# Function to cross-validate a GBM model with different numbers of trees, then
```

Comment:
In my stuff I estimate the optimum number of trees with `gbm.perf()`. I think it might be faster than doing cv manually.